pipeline/eval/eval.py

```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/env python3
+"""
+Evaluate a trained model with both the BLEU and chrF metrics.
+
+Kinds:
+   taskcluster/kinds/evaluate/kind.yml
+   taskcluster/kinds/evaluate-quantized/kind.yml
+   taskcluster/kinds/evaluate-teacher-ensemble/kind.yml
+
+Example usage:
+
+    $VCS_PATH/pipeline/eval/eval.py
+        --src               en                                                 \\
+        --trg               ca                                                 \\
+        --marian_config     fetches/final.model.npz.best-chrf.npz.decoder.yml  \\
+        --models            fetches/final.model.npz.best-chrf.npz              \\
+        --dataset_prefix    fetches/wmt09                                      \\
+        --artifacts_prefix  artifacts/wmt09                                    \\
+        --model_variant     gpu                                                \\
+        --workspace         12000                                              \\
+        --gpus              4
+
+Artifacts:
+
+For instance for a artifacts_prefix of: "artifacts/wmt09":
+
+  artifacts
+  ├── wmt09.en             The source sentences
+  ├── wmt09.ca             The target output
+  ├── wmt09.ca.ref         The original target sentences
+  ├── wmt09.log            The Marian log
+  ├── wmt09.metrics        The BLEU and chrF score
+  └── wmt09.metrics.json   The BLEU and chrF score in json format
+
+Fetches:
+
+For instance for a value of: "fetches/wmt09":
+  fetches
+  ├── wmt09.en.zst
+  └── wmt09.ca.zst
+"""
+
+
+import argparse
+import json
+import os
+import subprocess
+from textwrap import dedent, indent
+from typing import Optional
+
+
+def run_bash_oneliner(command: str):
+    """
+    Runs multi-line bash with comments as a one-line command.
+    """
+    command_dedented = dedent(command)
+
+    # Remove comments and whitespace.
+    lines = [
+        line.strip() for line in command_dedented.split("\n") if line and not line.startswith("#")
+    ]
+    command = " \\\n".join(lines)
+
+    print("-----------------Running bash in one line--------------")
```

Comment:
This should eventually be put into a shared location like `pipeline/utils`.

Reply:
Here is some example output:

```
[task 2024-01-30T15:51:31.305Z] The eval script is configured with the following:
[task 2024-01-30T15:51:31.305Z]  >          artifacts_dir: /home/ubuntu/tasks/task_170662979767690/artifacts
[task 2024-01-30T15:51:31.305Z]  > source_file_compressed: /home/ubuntu/tasks/task_170662979767690/fetches/devtest.en.zst
[task 2024-01-30T15:51:31.305Z]  >            source_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.en
[task 2024-01-30T15:51:31.305Z]  >            target_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru
[task 2024-01-30T15:51:31.305Z]  >        target_ref_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru.ref
[task 2024-01-30T15:51:31.305Z]  >         marian_decoder: "$MOZ_FETCHES_DIR"/marian-decoder
[task 2024-01-30T15:51:31.305Z]  >        marian_log_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.log
[task 2024-01-30T15:51:31.305Z]  >          language_pair: en-ru
[task 2024-01-30T15:51:31.305Z]  >           metrics_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics
[task 2024-01-30T15:51:31.305Z]  >           metrics_json: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics.json
[task 2024-01-30T15:51:31.305Z]  >      marian_extra_args: ['--workspace', '12000', '--devices', '0']
[task 2024-01-30T15:51:31.305Z] Ensure that the artifacts directory exists.
[task 2024-01-30T15:51:31.305Z] Save the original target sentences to the artifacts
[task 2024-01-30T15:51:31.305Z] -----------------Running bash in one line--------------
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   zstdmt -dc "/home/ubuntu/tasks/task_170662979767690/fetches/devtest.ru.zst" > "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru.ref"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z] -------------------------------------------------------
[task 2024-01-30T15:51:31.305Z] -----------------Running bash in one line--------------
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Decompress the source file, e.g. $fetches/wmt09.en.gz
[task 2024-01-30T15:51:31.305Z]   zstdmt -dc "/home/ubuntu/tasks/task_170662979767690/fetches/devtest.en.zst"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Tee the source file into the artifacts directory, e.g. $artifacts/wmt09.en
[task 2024-01-30T15:51:31.305Z]   | tee "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.en"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Take the source and pipe it in to be decoded (translated) by Marian.
[task 2024-01-30T15:51:31.305Z]   | "$MOZ_FETCHES_DIR"/marian-decoder
[task 2024-01-30T15:51:31.305Z]       --models /home/ubuntu/tasks/task_170662979767690/fetches/final.model.npz.best-chrf.npz
[task 2024-01-30T15:51:31.305Z]       --config /home/ubuntu/tasks/task_170662979767690/fetches/final.model.npz.best-chrf.npz.decoder.yml
[task 2024-01-30T15:51:31.305Z]       --quiet
[task 2024-01-30T15:51:31.305Z]       --quiet-translation
[task 2024-01-30T15:51:31.305Z]       --log /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.log
[task 2024-01-30T15:51:31.305Z]       --workspace 12000 --devices 0
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # The translations be "tee"ed out to the artifacts, e.g. $artifacts/wmt09.ca
[task 2024-01-30T15:51:31.305Z]   | tee "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Take in the translations via stdin and run the evaluation using sacrebleu.
[task 2024-01-30T15:51:31.305Z]   # The translations will be compared to the original corpus via the target_ref_file
[task 2024-01-30T15:51:31.305Z]   | sacrebleu "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru.ref"
[task 2024-01-30T15:51:31.305Z]       --language-pair en-ru
[task 2024-01-30T15:51:31.305Z]       --metrics bleu chrf
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Finally, tee out the metrics to the json file.
[task 2024-01-30T15:51:31.305Z]   | tee "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics.json"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z] -------------------------------------------------------
[task 2024-01-30T15:51:31.305Z] Look up the bleu and chrf from the output files from: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics.json
[task 2024-01-30T15:51:31.305Z] Metrics:  [{'name': 'BLEU', 'score': 0.0, 'signature': 'nrefs:1|case:mixed|eff:no|tok:13a|smooth:exp|version:2.3.1', 'verbose_score': '49.8/0.0/8.3/0.0 (BP = 0.000 ratio = 0.087 hyp_len = 2027 ref_len = 23295)', 'nrefs': '1', 'case': 'mixed', 'eff': 'no', 'tok': '13a', 'smooth': 'exp', 'version': '2.3.1'}, {'name': 'chrF2', 'score': 0.5, 'signature': 'nrefs:1|case:mixed|eff:yes|nc:6|nw:0|space:no|version:2.3.1', 'nrefs': '1', 'case': 'mixed', 'eff': 'yes', 'nc': '6', 'nw': '0', 'space': 'no', 'version': '2.3.1'}]
[task 2024-01-30T15:51:31.305Z] Saving the metrics in the older "text" format: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics
```

Reply:
thanks for logging marian arguments, it's important to see what's happening